java-frontend/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -417,4 +419,37 @@ public SymbolicValue exitValue() {
   public boolean exitingOnRuntimeException() {
     return exitSymbolicValue instanceof SymbolicValue.ExceptionalSymbolicValue && ((SymbolicValue.ExceptionalSymbolicValue) exitSymbolicValue).exceptionType() == null;
   }
+
+  Set<LearnedConstraint> learnedConstraints(ProgramState parent) {
+    ImmutableSet.Builder<LearnedConstraint> result = ImmutableSet.builder();
+    constraints.forEach((sv, childConstraint) -> {
+      Constraint parentConstraint = parent.getConstraint(sv);
+      if (parentConstraint == null || !parentConstraint.equals(childConstraint)) {
+        addLearnedConstraint(result, sv, childConstraint);
+      }
+    });
+    return result.build();
+  }
+
+  private static void addLearnedConstraint(ImmutableSet.Builder<LearnedConstraint> result, SymbolicValue sv, Constraint c) {
+    result.add(new LearnedConstraint(sv, c));
+    // FIXME this might end up adding twice the same SV in learned constraints. Safe because null constraints are filtered anyway
+    // FIXME remove null usage
+    if (sv instanceof BinarySymbolicValue) {
+      BinarySymbolicValue binarySymbolicValue = (BinarySymbolicValue) sv;
+      addLearnedConstraint(result, binarySymbolicValue.getLeftOp(), null);
+      addLearnedConstraint(result, binarySymbolicValue.getRightOp(), null);
+    }
+  }
+
+  Set<LearnedAssociation> learnedAssociations(ProgramState parent) {
+    ImmutableSet.Builder<LearnedAssociation> result = ImmutableSet.builder();
+    values.forEach((s, sv) -> {
+      if (parent.getValue(s) != sv) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)